misk/src/main/kotlin/misk/ServiceGraphBuilder.kt

```diff
@@ -0,0 +1,133 @@
+package misk
+
+import com.google.common.collect.HashMultimap
+import com.google.common.collect.LinkedHashMultimap
+import com.google.common.collect.Multimap
+import com.google.common.collect.SetMultimap
+import com.google.common.util.concurrent.Service
+import com.google.common.util.concurrent.ServiceManager
+import com.google.inject.Key
+import misk.CoordinatedService2.Companion.CycleValidity
+
+/**
+ * Builds a graph of `CoordinatedService2`s which defer start up and shut down until their dependent
+ * services are ready.
+ */
+class ServiceGraphBuilder {
+
+  private var serviceMap = mutableMapOf<Key<*>, CoordinatedService2>() // map of all services
+  private val dependencyMap = LinkedHashMultimap.create<Key<*>, Key<*>>()
+  private val enhancementMap = LinkedHashMultimap.create<Key<*>, Key<*>>()
```

Comment:
should this be a regular map? No duplicate keys right?

Reply:
(and if you get multiple puts on the same key, you can fail on the 2nd put, not later in a validation step)

Reply:
I could do this if I swap the mapping.

Right now it's _service_ to _set of enhancements_, and the validation works by making sure no enhancement is in multiple sets.

Reply:
Alright this has been updated in commit 5b34926. Note that _`ServiceGraphBuilder.addEnhancement()`_ now throws from a check if there is an attempt to apply an enhancement to more than one service.

I also modified the _CoordinatedService2_ API to take varargs for `addDependencies()` and `addEnhancements()` respectively, since as a consequence of this change we are no longer always adding a list of services.